### PR TITLE
Feature split dockerimage

### LINKED ISSUE #222: [Enhancement] Split docker image build into base image and application image
For the moment, the job build-docker-image in ci.yml takes some hours. 
### The reason:
- The whole container is rebuild everytime it runs
- The 2 components jbig2enc and pikepdf ( only arm ) get rebuild for every new build, though there're no changes

### A possible Solution:
Seperate the docker build into 2 stages:
- Build the base image with all deps only when there has been changes
- Base the final image on the base image

I've started a POC in my fork and will report if there's any progress

Regards

Schnuffle

### LINKED ISSUE #266: [BUG] Authentication requires libpq version 10 or above
**Describe the bug**
The following error is reported when starting paperless on my Raspberry Pi:
`ERROR SCRAM authentication requires libpq version 10 or above`
This results in a Server Error 500 after logging in.

**To Reproduce**
1. `docker-compose.yml`:
```
version: "3.4"

networks:
  default:
    external: true
    name: ${CADDY_NETWORK}

services:
  broker:
    image: redis:6.0
    volumes:
      - ./redis:/data
    restart: unless-stopped

  db:
    image: postgres:13
    volumes:
      - ./postgres:/var/lib/postgresql/data
    environment:
      POSTGRES_DB: paperless
      POSTGRES_USER: paperless
      POSTGRES_PASSWORD: paperless
    restart: unless-stopped

  paperless:
    image: ghcr.io/paperless-ngx/paperless-ngx:latest
    container_name: ${CONTAINER_NAME_PAPERLESS}
    environment:
      PAPERLESS_REDIS: redis://broker:6379
      PAPERLESS_DBHOST: db
      PAPERLESS_TIKA_ENABLED: 1
      PAPERLESS_TIKA_GOTENBERG_ENDPOINT: http://gotenberg:3000
      PAPERLESS_TIKA_ENDPOINT: http://tika:9998
      USERMAP_UID: 1001
      USERMAP_GID: 1001
      PAPERLESS_TIME_ZONE: ${TZ}
    volumes:
      - ./data:/usr/src/paperless/data
      - ./media:/usr/src/paperless/media
      - ./export:/usr/src/paperless/export
      - ./consume:/usr/src/paperless/consume
    restart: unless-stopped
    depends_on:
      - db
      - broker
      - gotenberg
      - tika

  gotenberg:
    image: gotenberg/gotenberg:7
    environment:
      CHROMIUM_DISABLE_ROUTES: 1
    restart: unless-stopped

  tika:
    image: iwishiwasaneagle/apache-tika-arm
    restart: unless-stopped
```
2. `docker compose run --rm paperless createsuperuser`
3. `docker compose down && docker compose up -d`

**Expected behavior**
Paperless should start correctly.

**Webserver logs**
```
12:19:59 [Q] ERROR Failed [louisiana-uncle-magnesium-eleven] - SCRAM authentication requires libpq version 10 or above
 : Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/base/base.py", line 219, in ensure_connection
    self.connect()
  File "/usr/local/lib/python3.9/site-packages/django/utils/asyncio.py", line 33, in inner
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/base/base.py", line 200, in connect
    self.connection = self.get_new_connection(conn_params)
  File "/usr/local/lib/python3.9/site-packages/django/utils/asyncio.py", line 33, in inner
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/postgresql/base.py", line 187, in get_new_connection
    connection = Database.connect(**conn_params)
  File "/usr/local/lib/python3.9/site-packages/psycopg2/__init__.py", line 122, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
psycopg2.OperationalError: SCRAM authentication requires libpq version 10 or above


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/django_q/cluster.py", line 432, in worker
    res = f(*task["args"], **task["kwargs"])
  File "/usr/src/paperless/src/paperless_mail/tasks.py", line 12, in process_mail_accounts
    for account in MailAccount.objects.all():
  File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 280, in __iter__
    self._fetch_all()
  File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 1324, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 51, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1173, in execute_sql
    cursor = self.connection.cursor()
  File "/usr/local/lib/python3.9/site-packages/django/utils/asyncio.py", line 33, in inner
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/base/base.py", line 259, in cursor
    return self._cursor()
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/base/base.py", line 235, in _cursor
    self.ensure_connection()
  File "/usr/local/lib/python3.9/site-packages/django/utils/asyncio.py", line 33, in inner
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/base/base.py", line 219, in ensure_connection
    self.connect()
  File "/usr/local/lib/python3.9/site-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/base/base.py", line 219, in ensure_connection
    self.connect()
  File "/usr/local/lib/python3.9/site-packages/django/utils/asyncio.py", line 33, in inner
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/base/base.py", line 200, in connect
    self.connection = self.get_new_connection(conn_params)
  File "/usr/local/lib/python3.9/site-packages/django/utils/asyncio.py", line 33, in inner
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/db/backends/postgresql/base.py", line 187, in get_new_connection
    connection = Database.connect(**conn_params)
  File "/usr/local/lib/python3.9/site-packages/psycopg2/__init__.py", line 122, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
django.db.utils.OperationalError: SCRAM authentication requires libpq version 10 or above


12:20:00 [Q] INFO recycled worker Process-1:2
12:20:00 [Q] INFO Process-1:6 ready for work at 85
12:21:00 [Q] ERROR SCRAM authentication requires libpq version 10 or above
```

**Relevant information**
 - Host OS of the machine running paperless: Raspberry Pi OS (64bit)
 - Version: 1.6.0
 - Installation method: docker

### LINKED ISSUE #451: Use the meta action to handle the creation of the standard labels.
> For the meta data use the meta action to handle the updates and creation of the standard labels. Makes it simpler. Also use the variables for the account instead of your name hard coded in the file. 

_Originally posted by @sytone in https://github.com/paperless-ngx/paperless-ngx/issues/238#issuecomment-1072946490_

Creating this issue since it's a good change and I don't want to forget it in the shuffle

### LINKED ISSUE #513: [BUG] compiled qpdf missing in amd64 docker image
**Describe the bug**

**To Reproduce**
pre consmption script with qpdf gives error command not found

**Expected behavior**
script being able to run qpdf to process pdf files

**Screenshots**

**Relevant information**
 - Synology Intel 
 - Browser [Edge]
 - Version [e.g. 1.6.0]
 - Installation method: [docker]
 - qpdf still present in arm builds
 - qpdf was present in paperless-ng

### PR DESCRIPTION
**Pull request by schnuffle**
This PR needs some work to be properly integrated into the default workflow.

For the moment I've defined 2 new workflows:
- Build a base image 
- Build a application image, referencing the base image

For that I also adopted the Dockerfile and created a new one for the base image.

During my testing I had to reallize that building the frontend also takes quite some time. My first idea has been to also create a base image for the frontend build. But that wasn't a good idea as now it would also build multiarch images ( so building 3 times, instead of  only once ) and the build is a lot slower in a docker environment.

My second idea has been to build the frontend not in docker but straight on the runner by using the action/setup-node.

So right now the whole build uses one runner:
- Prepare the runner for nodejs and build frontend
- Prepare runner for docker and create image and push to registry

Open questions:
- How should we trigger new base image creations?
- Do we need to create different base images for dev and other branches?
- To be able to push to the org registry we need to use a token as the default can't be used, at least not on my fork.

### CODE REVIEW

**Comment by qcasey** [comment]
Thanks @schnuffle, this looks great!

> How should we trigger new base image creations?

Could we do this [manually](https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/) and when the Docker file changes? 

> Do we need to create different base images for dev and other branches?

I think this is a yes. Tagging the base image with :dev would be ideal.

> To be able to push to the org registry we need to use a token as the default can't be used, at least not on my fork.

This will work as-is when merged into the -ngx repo

**Comment by qcasey** [comment]
Do you think it's possible to push these build containers to another repo? Such as paperless-ngx/paperless-ngx-builder ?

**Comment by schnuffle** [comment]
>Could we do this [manually](https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/) and when the Docker file changes?

The on push directive offers paths, tags and branch filters. I read that using combined branch and paths filters results in both filters being combined ( trigger when branch and paths ).
That would allow creating base images on file change and for distinct branches. And with job if: expressions you can further cutomize.

I'll take a look at workflow_dispatch, might have saved a lot of tag pushing during test :)
> I think this is a yes. Tagging the base image with :dev would be ideal.

Okay, I think that's not a problem, just need to think about how to create the tags in the prepare step. For the moment it's build when pushing a tag like ngx-base0.0.1 which creates the tags 0.0.1 and latest.
Should we create to different base images, like one ngx-base and one ngx-base-dev?
Because only having a :dev tag would mean you can only have one version. With a dev image you could create base image for py3.9, py3.10, different requirements ........

Regards
Schnuffle

**Comment by schnuffle** [comment]
And to have yet another approach.
How about createingcthe pikepdf wheel?
Is the a pypy repo on github?

**Comment by qcasey** [comment]
> And to have yet another approach.
How about createing the pikepdf wheel?
Is there a pypy repo on github?

That also very interesting, could it be uploaded to github packages?

**Comment by qcasey** [comment]
I'm out of my depth here, so however you'd like to proceed we can test with. I really appreciate all your efforts here.

**Comment by schnuffle** [comment]
I've progressed on my latest idea and combined the 2 approaches:

1. Build wheels in docker for armv7/64 and push to registry
2. Extract those wheels and push to release package. I also create wheels of the dependencies LXML and Pillow as they would need to be compiled as well
3. Create base image by installing the created wheels without the need to compile
4. Build the application image from the base image

For that approach I created a repo for wheel building: https://github.com/schnuffle/build-pikepdfwheel that offers the assets as release.

The different stages do not have proper workflow trigger yet and the whole process needs manual interaction. It hasn't made sense till now as I needed to run quite some builds to straighten out errors and problems.

Looking forward to some input.

One low hanging fruit is still available, which is building jbig2enc and make a release, so it can just be pulled in instead of compiling it every time.

To save time I dropped amd64 for the moment, readding it needs only some minor changes to the Dockerfile

Regards

Schnuffle

**Comment by schnuffle** [comment]
Most activity for the moment is done in https://github.com/schnuffle/build-pikepdfwheel

As the builds take ages, I'm starting to split into more and more finer images to be able to shorten the time for the next step.

**Comment by sytone** [comment]
For the meta data use the meta action to handle the updates and creation of the standard labels. Makes it simpler. Also use the variables for the account instead of your name hard coded in the file. 

@shamoon as a FYI as you are talking about docker. Also I noticed that linuxserver.oi just had a repo added for ngx

File: .github/workflows/build-project.yml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      -
+        name: Prepare Docker
+        id: prepare
+        run: |
+          IMAGE_NAME=ghcr.io/schnuffle/paperless-docker
+          TAGS=${IMAGE_NAME}:latest
+          INSPECT_TAG=${IMAGE_NAME}:latest
+          echo ::set-output name=tags::${TAGS}
+          echo ::set-output name=inspect_tag::${INSPECT_TAG}
+      -
```
Comment (sytone) [inline .github/workflows/build-project.yml:32]: Add the metadata-action here. I am assuming we are using semver. 

```
      - name: Docker meta
        id: meta
        uses: docker/metadata-action@v3
        with:
          images: ghcr.io/${{ github.repository_owner }}/paperless-docker
          tags: |
            type=ref,event=branch
            type=ref,event=pr
            type=semver,pattern={{version}}
            type=semver,pattern={{major}}.{{minor}}
```

File: .github/workflows/build-project.yml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+          password: ${{ secrets.GITHUB_TOKEN }}
+      - 
+        name: Build and push
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          file: ./Dockerfile
+          platforms: linux/amd64,linux/arm/v7,linux/arm64
+          push: true
+          tags: ${{ steps.prepare.outputs.tags }}
```
Comment (sytone) [inline .github/workflows/build-project.yml:66]: Replace tags with

```
          tags: ${{ steps.meta.outputs.tags }}
          labels: ${{ steps.meta.outputs.labels }}
```

File: .github/workflows/docker-wheel.yml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      -
+        name: Prepare
+        id: prepare
+        run: |
+          IMAGE_NAME=ghcr.io/schnuffle/paperless-ngx-basewheel
+          TAGS=${IMAGE_NAME}:latest
+          INSPECT_TAG=${IMAGE_NAME}:latest
+          echo ::set-output name=tags::${TAGS}
+          echo ::set-output name=inspect_tag::${INSPECT_TAG}
+      -
```
Comment (sytone) [inline .github/workflows/docker-wheel.yml:21]: Add

```
      - name: Docker meta
        id: meta
        uses: docker/metadata-action@v3
        with:
          images: ghcr.io/${{ github.repository_owner }}/paperless-docker
          tags: |
            type=ref,event=branch
            type=ref,event=pr
            type=semver,pattern={{version}}
            type=semver,pattern={{major}}.{{minor}}
```

File: .github/workflows/docker-wheel.yml [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+          password: ${{ secrets.GITHUB_TOKEN }}
+      -
+        name: Build and push
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          file: ./Dockerfile.base-ngxwheel
+          platforms: linux/arm/v7,linux/arm64
+          push: true
+          tags: ${{ steps.prepare.outputs.tags }}
```
Comment (sytone) [inline .github/workflows/docker-wheel.yml:45]: replace tags with

```
          tags: ${{ steps.meta.outputs.tags }}
          labels: ${{ steps.meta.outputs.labels }}
```

File: .github/workflows/build-project.yml (hunk 1 above)
Comment (qcasey) [reply .github/workflows/build-project.yml:32]: This is a good note but I think it'd be best to put this in a new PR to minimize the diff here.

**Comment by qcasey** [comment]
Nice, getting close.

It's difficult to determine what's changing because of the new file. This is how I suggest testing, I'm open to suggestions:

1. Rename ci2.yml to ci.yml - we'd be replacing anyway right?
2. Move Docker.frombase to the .github directory since it's only really used in CI.
3. This push should trigger CI to build another `feature-split-dockerimage` container we can test to ensure functionality

**Comment by qcasey** [comment]
Also `build-project.yml` is no longer needed, right?

**Comment by schnuffle** [comment]
- build-project.yml is not needed anymore
- The ci2.yml is the original workflow adopted to use the base image. 
- I moved my Dockerfile to Dockfile.frombase to have both available.

@sytone I'll have to look into the labels and tags, thanxs for the feedback

Regards
Schnuffle

File: Dockerfile.frombase [hunk 5]
```diff
@@ -0,0 +1,48 @@
+FROM ghcr.io/paperless-bgx/ngx-builder:latest
```
Comment (stumpylog) [inline Dockerfile.frombase:1]: `bgx` should be `ngx`

**Comment by stumpylog** [comment]
Before I get too far down a rabbit hole with the base image, what are people's thoughts on locking the `pikepdf` and `psycopg2` versions?  It's maybe something discussed already, but I missed if it was.

If we set the following versions, we can instead use the system provided packages from backports.
- `pikepdf == 5.0.1`
- `psycopg2 == 2.9.2`

Advantages:
- Faster image build times
- Building can be done in 1 repo
- Package versions don't need to be synced to a different repo

Disadvantages:
- Behind on versions
- Need to check Debian package versions and update our version on updates there

**Comment by qcasey** [comment]
That sounds fine too if it works. There is a new CI error, I think it's a dependency problem building qpdf from scratch:

https://github.com/paperless-ngx/paperless-ngx/runs/5689070850?check_suite_focus=true

If that's true, all the more reason to lock something in, either a dpkg or prebuilt image.

**Comment by shamoon** [comment]
Tracking https://github.com/pikepdf/pikepdf/pull/324

**Comment by shamoon** [comment]
We're waiting with bated breath! =)

qcasey already approved the builder PR, you good with merging?

ps. pikepdf 5.1.0 should have fixed that project.toml thing yea?

**Comment by stumpylog** [comment]
I know qcasey had an initial look, but I don't think its approved yet.  Approval requirements might not be setup on that repo?  Although I have the power to merge it, more people should probably check my work a bit.

pikepdf v5.1.1 says it fixed the issue, but for the moment we're on v5.1.0.  I'd prefer to stay there while finalizing the images.

**Comment by qcasey** [comment]
We can wait for a few more eyes on the builder PR if you'd like, although I imagine most of the hands on testing would be done here in this PR

**Comment by gador** [comment]
I was wondering why https://github.com/paperless-ngx/builder/pull/1 was merged into `main` and this PR targets `dev`?
Doesn't that make testing a bit harder, because one would have to merge (or rebase) `main` first into `dev`?

**Review by gador — CHANGES_REQUESTED** [review]
When using the Dockerfile.frombase as the dockerfile for my docker compose, it does build and works as expected.
(Current change to `dev` makes it build and work :+1: )

File: Dockerfile.frombase (hunk 5 above)
[the remark this answers is not in the document]
Comment (stumpylog) [reply Dockerfile.frombase:1]: This will be fixed up once the builder 1.0.0 tag finishes

File: .github/workflows/build-project.yml (hunk 1 above)
Comment (stumpylog) [reply .github/workflows/build-project.yml]: I've actually included the `metadata-action`, but please check my attempts to generate the right values and such from tags and branches

File: .github/workflows/build-project.yml (hunk 2 above)
Comment (stumpylog) [reply .github/workflows/build-project.yml]: This is now updated in the `ci.yml`

File: .github/workflows/docker-wheel.yml (hunk 3 above)
Comment (stumpylog) [reply .github/workflows/docker-wheel.yml]: I've updated the `ci.yml` to roughly do this

File: .github/workflows/docker-wheel.yml (hunk 4 above)
Comment (stumpylog) [reply .github/workflows/docker-wheel.yml]: Done now in `ci.yml`

File: Dockerfile.frombase (hunk 5 above)
[the remark this answers is not in the document]
Comment (stumpylog) [reply Dockerfile.frombase]: Fixed now

**Comment by stumpylog** [comment]
Alright, this pull request is now ready.  I've locally built an image and started it up without issue.

As per the Dockerfile, it uses the base image, 1.0 (so we'll get any bugfixes from there, but not new dependency versions).  That base image includes:
| What | Version |
| ------ | -------- |
| jbig2enc | 0.29 |
| qpdf | 10.6.2 |
| pikepdf | 5.1.0 |
| psycopg2 | 2.9.3 |

in addition to the other required packages, like tesseract, optipng, etc.

So please, kick the tires a lot and lets see if anything breaks,

**Comment by shamoon** [comment]
@schnuffle @stumpylog are we ready to go for it?? Not that I’m eager or anything… 😝

**Comment by stumpylog** [comment]
I'm good with this as an initial version.  The base image 1.0.1 to fix `pg_ready` has just finished building and my local test stack with it seemed to be doing just fine (though now re-building is affected by the npm build issue).

The only thing I can think of it about how the image labeling works.  But that's kind of hard to test without creating a bunch of tags and can be tested/fixed if need be when we create the next release.

**Comment by qcasey** [comment]
>  But that's kind of hard to test without creating a bunch of tags and can be tested/fixed if need be when we create the next release.

That's what we did last release, we were fixing the plane while taking off.

**Comment by shamoon** [comment]
> That's what we did last release, we were fixing the plane while taking off. The next beta will be a good test run.

Haha exactly. You know how I feel about this—that's why they call it a `dev` branch =)

Maybe we merge #610 to make sure thats out of the way then youre free to merge this and test / tweak to your hearts content?

**Comment by shamoon** [comment]
Yay https://github.com/paperless-ngx/paperless-ngx/actions/runs/2080985572

I say go for it!

**Review by qcasey — APPROVED** [review]

**Comment by shamoon** [comment]
Bombs away! Bravo on this one guys, what a mountain of work (I realize probably will be some more =)

Btw, I recently found this and its awesome though maybe this is beyond in complexity https://github.com/nektos/act

**Comment by shamoon** [comment]
Let the tweaking begin! https://github.com/paperless-ngx/paperless-ngx/runs/5798588214?check_suite_focus=true
😝

**Comment by qcasey** [comment]
I tried `act` a few weeks ago for this project and didn't have any luck, although I don't remember quite why. 

Anyway, failing in just 5s is a new record for us at least 🚀

I feel bad for only commenting from the sidelines but I'll be back in this coming week to help fix these things. It will be great to fix this and draft a new beta. Have a good weekend!

**Pull request merged**